### PR TITLE
Task#252 refresh btn

### PR DESCRIPTION
**Pull request by NemTechCN**

### CODE REVIEW

File: src/components/check-password-dialog/CheckPasswordDialogTs.ts [hunk 1]
```diff
@@ -19,6 +19,9 @@ export class CheckPasswordDialogTs extends Vue {
     @Prop()
     showCheckPWDialog: boolean
 
+    @Prop({default: ''})
```
Comment (decentraliser) [inline src/components/check-password-dialog/CheckPasswordDialogTs.ts]: what is formItems doing in this componetn?

Comment (NemTechCN) [reply src/components/check-password-dialog/CheckPasswordDialogTs.ts]: it is useless,

Comment (NemTechCN) [reply src/components/check-password-dialog/CheckPasswordDialogTs.ts]: it is used to solve another issue，

Comment (NemTechCN) [reply src/components/check-password-dialog/CheckPasswordDialogTs.ts]: This line of code has been deleted

File: src/views/mosaic/mosaic-list/MosaicList.vue [hunk 2]
```diff
@@ -60,6 +59,8 @@
             <Icon v-if="mosaicSortType.byAlias == currentSortType" class="active_sort_type"
                   :type="sortDirection?'md-arrow-dropdown':'md-arrow-dropup'"/>
             </span>
+          <span @click="refreshMosaicList(wallet)"
```
Comment (decentraliser) [inline src/views/mosaic/mosaic-list/MosaicList.vue]: wallet is not needed here

File: src/views/mosaic/mosaic-list/MosaicListTs.ts [hunk 3]
```diff
@@ -142,6 +148,27 @@ export class MosaicListTs extends Vue {
...
+            initMosaic(wallet, this.$store)
+            this.$Notice.destroy()
+            this.$Notice.success({title: '' + this.$t(Message.SUCCESS)})
+            this.$store.commit('SET_MOSAICS_LOADING', false)
+        } catch (error) {
+            console.error("App -> refresh -> error", error)
+            this.$store.commit('SET_MOSAICS_LOADING', false)
+        }
+    }
+
```
Comment (decentraliser) [inline src/views/mosaic/mosaic-list/MosaicListTs.ts]: Here, every first click after mounted() throws the REFRESH_TOO_FAST_WARNING error

File: src/views/mosaic/mosaic-list/MosaicListTs.ts [hunk 4]
```diff
@@ -142,6 +148,27 @@ export class MosaicListTs extends Vue {
...
+            this.$Notice.warning({title: '' + this.$t(Message.REFRESH_TOO_FAST_WARNING)})
+            return
+        }
+        this.$store.commit('SET_MOSAICS_LOADING', true)
+        this.mosaicRefreshTimestamp = currentTimestamp
+        try {
+            initMosaic(wallet, this.$store)
+            this.$Notice.destroy()
+            this.$Notice.success({title: '' + this.$t(Message.SUCCESS)})
+            this.$store.commit('SET_MOSAICS_LOADING', false)
```
Comment (decentraliser) [inline src/views/mosaic/mosaic-list/MosaicListTs.ts]: SET_MOSAICS_LOADING is managed in initMosaic, you can remove it from here

File: src/views/mosaic/mosaic-list/MosaicListTs.ts [hunk 5]
```diff
@@ -142,6 +148,27 @@ export class MosaicListTs extends Vue {
...
+        } catch (error) {
+            console.error("App -> refresh -> error", error)
```
Comment (decentraliser) [inline src/views/mosaic/mosaic-list/MosaicListTs.ts]: Naming console.error is nice, however, please use an exact description, eg. `refreshMosaicList -> error `

File: src/views/mosaic/mosaic-list/MosaicListTs.ts [hunk 6]
```diff
@@ -142,6 +148,27 @@ export class MosaicListTs extends Vue {
...
+            this.$store.commit('SET_MOSAICS_LOADING', false)
```
Comment (decentraliser) [inline src/views/mosaic/mosaic-list/MosaicListTs.ts]: SET_MOSAICS_LOADING is managed in initMosaic, you can remove it from here

File: src/views/namespace/namespace-function/namespace-list/NamespaceListTs.ts [hunk 7]
```diff
@@ -140,6 +146,25 @@ export class NamespaceListTs extends Vue {
...
+            this.$Notice.warning({title: '' + this.$t(Message.REFRESH_TOO_FAST_WARNING)})
+            return
+        }
+        this.$store.commit('SET_NAMESPACE_LOADING', true)
+        this.namespaceRefreshTimestamp = currentTimestamp
+        try {
+            getNamespacesFromAddress(wallet.address, node)
+            this.$store.commit('SET_NAMESPACE_LOADING', false)
+        } catch (error) {
+            console.error("App -> refresh -> error", error)
```
Comment (decentraliser) [inline src/views/namespace/namespace-function/namespace-list/NamespaceListTs.ts]: error description

File: src/views/namespace/namespace-function/namespace-list/NamespaceListTs.ts [hunk 8]
```diff
@@ -140,6 +146,25 @@ export class NamespaceListTs extends Vue {
         return ''
     }
 
+    async refreshNamesapceList() {
```
Comment (decentraliser) [inline src/views/namespace/namespace-function/namespace-list/NamespaceListTs.ts]: typo namespace

File: src/views/mosaic/mosaic-list/MosaicList.vue (hunk 2 above)
Comment (NemTechCN) [reply src/views/mosaic/mosaic-list/MosaicList.vue]: fixed

File: src/views/namespace/namespace-function/namespace-list/NamespaceListTs.ts [hunk 9]
```diff
@@ -140,6 +146,25 @@ export class NamespaceListTs extends Vue {
...
+        const currentTimestamp = new Date().valueOf()
+        if (currentTimestamp - namespaceRefreshTimestamp <= 2000) {
+            this.$Notice.destroy()
+            this.$Notice.warning({title: '' + this.$t(Message.REFRESH_TOO_FAST_WARNING)})
+            return
+        }
+        this.$store.commit('SET_NAMESPACE_LOADING', true)
+        this.namespaceRefreshTimestamp = currentTimestamp
+        try {
+            getNamespacesFromAddress(wallet.address, node)
```
Comment (decentraliser) [inline src/views/namespace/namespace-function/namespace-list/NamespaceListTs.ts]: here you are not refreshing the namespaces
getNamespacesFromAddress is returning a value, that you need to commit to the store afterwards

Please refer to this:
https://github.com/nemfoundation/nem2-desktop-wallet/blob/3864e257f3403cb3d069ee980b900550a5268f44/src/App.vue#L109-L116

File: src/views/mosaic/mosaic-list/MosaicListTs.ts (hunk 4 above)
Comment (NemTechCN) [reply src/views/mosaic/mosaic-list/MosaicListTs.ts]: fixed

File: src/views/mosaic/mosaic-list/MosaicListTs.ts (hunk 5 above)
Comment (NemTechCN) [reply src/views/mosaic/mosaic-list/MosaicListTs.ts]: fixed

File: src/views/mosaic/mosaic-list/MosaicListTs.ts (hunk 6 above)
Comment (NemTechCN) [reply src/views/mosaic/mosaic-list/MosaicListTs.ts]: fixed

File: src/views/namespace/namespace-function/namespace-list/NamespaceListTs.ts (hunk 7 above)
Comment (NemTechCN) [reply src/views/namespace/namespace-function/namespace-list/NamespaceListTs.ts]: fixed

File: src/views/namespace/namespace-function/namespace-list/NamespaceListTs.ts (hunk 8 above)
Comment (NemTechCN) [reply src/views/namespace/namespace-function/namespace-list/NamespaceListTs.ts]: fixed

File: src/views/namespace/namespace-function/namespace-list/NamespaceListTs.ts (hunk 9 above)
Comment (NemTechCN) [reply src/views/namespace/namespace-function/namespace-list/NamespaceListTs.ts]: fixed

File: src/views/namespace/namespace-function/namespace-list/NamespaceListTs.ts [hunk 10]
```diff
@@ -140,6 +146,25 @@ export class NamespaceListTs extends Vue {
...
+        const currentTimestamp = new Date().valueOf()
+        if (currentTimestamp - namespaceRefreshTimestamp <= 2000) {
+            this.$Notice.destroy()
+            this.$Notice.warning({title: '' + this.$t(Message.REFRESH_TOO_FAST_WARNING)})
+            return
+        }
+        try {
+            this.namespaceRefreshTimestamp = currentTimestamp
+            const initMosaicsAndNamespaces = await Promise.all([getNamespacesFromAddress(wallet.address, node)])
+            this.$store.commit('SET_NAMESPACES', initMosaicsAndNamespaces[0] || [])
```
Comment (decentraliser) [inline src/views/namespace/namespace-function/namespace-list/NamespaceListTs.ts]: you don't think that something like that would look a bit better?

```
const namespaces = await getNamespacesFromAddress(wallet.address, node)
this.$store.commit('SET_NAMESPACES', namespaces)
```

**Pull request merged**